output/sarif.go

```diff
@@ -0,0 +1,282 @@
+				WithLevel("warning").
+				WithMessage(sarif.NewTextMessage(warning.Message)).
+				AddLocation(
+					sarif.NewLocationWithPhysicalLocation(
+						sarif.NewPhysicalLocation().
+							WithArtifactLocation(
+								sarif.NewSimpleArtifactLocation(filepath.ToSlash(result.FileName)),
+							),
+					),
+				)
+		}
+
+		// Process exceptions
+		for _, exception := range result.Exceptions {
+			ruleID := getRuleID(result.Namespace, "allow")
+			var idx int
+			if existingIdx, ok := getRuleIndex(ruleID, indices); ok {
+				idx = existingIdx
+			} else {
+				idx = addRuleIndex(run, ruleID, exception, indices)
+			}
+
+			run.CreateResultForRule(ruleID).
+				WithRuleIndex(idx).
+				WithLevel("note").
+				WithMessage(sarif.NewTextMessage(exception.Message)).
+				AddLocation(
+					sarif.NewLocationWithPhysicalLocation(
+						sarif.NewPhysicalLocation().
+							WithArtifactLocation(
+								sarif.NewSimpleArtifactLocation(filepath.ToSlash(result.FileName)),
+							),
+					),
+				)
+		}
+
+		// Add success or skipped result if no other results
+		if len(result.Failures) == 0 && len(result.Warnings) == 0 && len(result.Exceptions) == 0 {
+			text := successDesc
+			ruleType := "success"
+			if result.Successes == 0 {
+				text = skippedDesc
+				ruleType = "skip"
+			}
+
+			emptyResult := Result{
+				Message: text,
+				Metadata: map[string]interface{}{
+					"description": text,
+				},
+			}
+			ruleID := getRuleID(result.Namespace, ruleType)
+			var idx int
+			if existingIdx, ok := getRuleIndex(ruleID, indices); ok {
+				idx = existingIdx
+			} else {
+				idx = addRuleIndex(run, ruleID, emptyResult, indices)
+			}
+
+			run.CreateResultForRule(ruleID).
+				WithRuleIndex(idx).
+				WithLevel("none").
+				WithMessage(sarif.NewTextMessage(text)).
+				AddLocation(
+					sarif.NewLocationWithPhysicalLocation(
+						sarif.NewPhysicalLocation().
+							WithArtifactLocation(
+								sarif.NewSimpleArtifactLocation(filepath.ToSlash(result.FileName)),
+							),
+					),
+				)
+		}
+	}
+
+	// Add run metadata
+	exitCode := 0
+	exitDesc := exitNoViolations
+	if hasFailures(results) {
+		exitCode = 1
+		exitDesc = exitViolations
+	} else if hasWarnings(results) {
+		exitDesc = exitWarnings
+	} else if hasExceptions(results) {
+		exitDesc = exitExceptions
+	}
+
+	successful := true
+	invocation := sarif.NewInvocation()
+	invocation.ExecutionSuccessful = &successful
+	invocation.ExitCode = &exitCode
+	invocation.ExitCodeDescription = &exitDesc
+
+	run.Invocations = []*sarif.Invocation{invocation}
+
+	// Add the run to the report
+	report.AddRun(run)
+
+	// Write the report
+	return report.Write(s.writer)
+}
+
+// Report is not supported in SARIF output
+func (s *SARIF) Report(_ []*tester.Result, _ string) error {
+	return fmt.Errorf("report is not supported in SARIF output")
+}
+
+// hasResults returns true if any of the results contain items in the specified field
+func hasResults(results []CheckResult, field string) bool {
```

Comment:
This defaults to false if there's a typo in the "field" parameter which could lead to unexpected behavior. Thinking about this more, I think these helpers should be attached to the CheckResult type directly. 

I sent https://github.com/open-policy-agent/conftest/pull/1063 for review to address this.

Reply:
No, let's not block your PR on that. We can revise later.